src/web100srv.c

```diff
@@ -1527,647 +1382,422 @@ void child_process(int parent_pipe, SSL_CTX *ssl_context, int ctlsockfd) {
+ * overload the server.
+ */
+int server_is_overloaded(int queue_size) {
+  if (queue) {
+    if (multiple) {
+      return queue_size >= 4 * max_clients;
+    } else {
+      return queue_size >= max_clients;
     }
+  } else {
+    return queue_size >= 1;
+  }
+}
 
-    if ((waiting == 0) && (head_ptr != NULL)) {
-      log_println(6, "Fault: Something [%d] in queue, but no waiting clients",
-                  head_ptr->pid);
-      while (head_ptr != NULL) {
-        /* send_msg(head_ptr->ctlsockfd, SRV_QUEUE, "9977", 4); */
-        child_signal = SRV_QUEUE_SERVER_BUSY;
-        write(head_ptr->pipe, &child_signal, sizeof(int));
-        close(head_ptr->pipe);
-        tpid = head_ptr->pid;
-        child_sig(-1);
-        kill(tpid, SIGTERM);
-        child_sig(tpid);
-      }
-      waiting = 0;
-      mclients = 0;
-      continue;
-    }
+/**
+ * Removes the element after okay_child from the list.  The calling function
+ * owns the memory containing the removed element.
+ * @param head The head of the list - may not be NULL or point to NULL
+ * @param okay_child The element before the element to remove - may be NULL
+ */
+void remove_next_child(ndtchild **head, ndtchild* okay_child) {
+  if (okay_child == NULL) {
+    // We are removing the head
+    *head = (*head)->next;
+  } else {
+    // We are removing from the middle of the list
+    okay_child->next = okay_child->next->next;
+  }
+}
 
-    if (head_ptr != NULL) {
-      if ((time(0) - head_ptr->stime) > 70) {  // 70 => WAIT-TIME-THRESH
-        log_println(6,
-                    "Fault: Something in queue, but child %d (fd=%d) has "
-                    "exceeded wait time",
-                    head_ptr->pid, head_ptr->pipe);
-        // Should send new 9977 'test aborted' signal to client.  Using
-        // this for now
-        // rac 3/26/10
-
-        log_println(6, "pid=%d, client='%s', stime=%ld, qtime=%ld now=%ld",
-                    head_ptr->pid, head_ptr->addr, head_ptr->stime,
-                    head_ptr->qtime, time(0));
-        log_println(6, "pipe-fd=%d, running=%d", head_ptr->pipe,
-                    head_ptr->running);
-        // boot the client
-        child_signal = SRV_QUEUE_SERVER_BUSY;
-        write(head_ptr->pipe, &child_signal, sizeof(int));
-        close(head_ptr->pipe);
-        tpid = head_ptr->pid;
-        child_sig(-1);
-        kill(tpid, SIGTERM);
-        child_sig(tpid);
-        continue;
-      }
-    }
+/**
+ * Enqueues a new child onto the list.
+ * @param head The head of the list
+ * @param new_child The child to enqueue
+ */
+void enqueue_child(ndtchild **head, ndtchild *new_child) {
+  ndtchild *current;
+  if (*head == NULL) {
+    // Enqueue onto the empty list
+    *head = new_child;
+  } else {
+    // Enqueue onto the end of a non-empty list
+    current = *head;
+    while (current->next != NULL) current = current->next;
+    current->next = new_child;
```

Comment:
Suggest `new_child->next = NULL;` I don't think we should make it the caller's responsibility to update the `next` pointer if they move an element from one queue to another